### PR TITLE
feat(data integration): Add Snowflake data integration

### PR DESCRIPTION
**Pull request by Meggielqk**

### CODE REVIEW

**Comment by Meggielqk** [comment]
Will add the Chinese file and update the architecture diagram later.

File: en_US/data-integration/snowflake.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+[Snowflake](https://www.snowflake.com/en/) is a cloud-based data platform that provides a highly scalable and flexible solution for data warehousing, analytics, and secure data sharing. Known for its ability to handle structured and semi-structured data, Snowflake is designed to store vast amounts of data while providing fast query performance and seamless integration with various tools and services.
+
+This page provides a detailed introduction to the data integration between EMQX and Snowflake, and offers practical guidance on the rule and Sink creation.
+
+## How It Works
+
+Snowflake data integration in EMQX is a ready-to-use feature that can be easily configured for complex business development. In a typical IoT application, EMQX acts as the IoT platform responsible for device connectivity and message transmission, while Snowflake serves as the data storage and processing platform, handling the ingestion, storage, and analysis of this message data.
+
+![azure-blob-storage-architecture](/Users/emqx/Documents/GitHub/emqx-docs/en_US/data-integration/assets/azure-blob-storage-architecture.png)
```
Comment (id) [inline en_US/data-integration/snowflake.md]: This does not look right.

File: en_US/data-integration/snowflake.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+EMQX utilizes rules engines and Sinks to forward device events and data to Snowflake. Applications can read data from Snowflake for further data applications. The specific workflow is as follows:
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
EMQX utilizes rules engines and Sinks to forward device events and data to Snowflake. End users and applications can then access data in Snowflake tables. The specific workflow is as follows:
```

File: en_US/data-integration/snowflake.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+1. **Device Connection to EMQX**: IoT devices trigger an online event upon successfully connecting via the MQTT protocol. The event includes device ID, source IP address, and other property information.
+2. **Device Message Publishing and Receiving**: Devices publish telemetry and status data through specific topics. EMQX receives the messages and compares them within the rules engine.
+3. **Rules Engine Processing Messages**: The built-in rules engine processes messages and events from specific sources based on topic matching. It matches corresponding rules and processes messages and events, such as data format transformation, filtering specific information, or enriching messages with context information.
+4. **Writing to Snowflake**: The rule triggers an action to write the message to Storage Container. Using the Snowflake Sink, users can extract the necessary data and store it in Snowflake in structured formats such as CSV files, ensuring the data is organized for efficient analysis.
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
4. **Writing to Snowflake**: The rule triggers an action to write messages to Snowflake Stage, and load it into a Snowflake table.
```

File: en_US/data-integration/snowflake.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+This section introduces the preparations required before creating a Snowflake Sink in EMQX.
+
+### Prerequisites
+
+- Understanding of [rules](./rules.md).
+- Understanding of [data integration](./data-bridges.md).
+
+### Initialize Snowflake ODBC driver
+
+To enable EMQX to communicate with Snowflake and efficiently transfer data, it is necessary to install and configure the Snowflake Open Database Connectivity (ODBC) driver. It acts as the communication bridge, ensuring that data is properly formatted, authenticated, and transferred.
```
Comment (id) [inline en_US/data-integration/snowflake.md:53]: Perhaps add a link to the official ODBC driver page https://docs.snowflake.com/en/developer-guide/odbc/odbc-download, and the license agreement https://sfc-repo.snowflakecomputing.com/odbc/Snowflake_ODBC_Driver_License_Agreement.pdf

File: en_US/data-integration/snowflake.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+## Create a Connector
+
+Before adding the Snowflake Sink, you need to create the corresponding connector in EMQX to establish the connection with Snowflake.
+
+1. Go to the Dashboard **Integration** -> **Connector** page.
+2. Click the **Create** button in the top right corner.
+3. Select **Snowflake** as the connector type and click next.
+4. Enter the connector name, a combination of upper and lowercase letters and numbers. Here, enter `my-snowflake`.
+5. Enter the connection information.
+   - **Server Host**: The server host is the Snowflake endpoint URL, typically in the format `<Your Snowflake Account>.snowflakecomputing.com`. You need to replace `<Your Snowflake Account>` with the subdomain specific to your Snowflake instance.
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
   - **Server Host**: The server host is the Snowflake endpoint URL, typically in the format `<Your Snowflake Organization ID>-<Your Snowflake Account>.snowflakecomputing.com`. You need to replace `<Your Snowflake Organization ID>-<Your Snowflake Account>` with the subdomain specific to your Snowflake instance.
```

File: en_US/data-integration/snowflake.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+   - **Account**: Enter your Snowflake account name, which is part of the URL you use to access the Snowflake platform and can be found in your Snowflake console.
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
   - **Account**: Enter your Snowflake Organization ID and Snowflake account name separated by a dash (`-`), which is part of the URL you use to access the Snowflake platform and can be found in your Snowflake console.
```

File: en_US/data-integration/snowflake.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+6. Select the `my-snowflake` connector created earlier from the connector dropdown. You can also click the create button next to the dropdown to quickly create a new connector in the pop-up box. The required configuration parameters can be found in [Create a Connector](#create-a-connector).
+
+7. Configure the following settings:
+
+   - **Database Name**: Enter `testdatabase`. This is the Snowflake database that was created for storing EMQX data.
+   - **Schema**: Enter `public`, the schema within the `testdatabase` where the data table is located.
+   - **Stage**: Enter `emqx`, the stage created in Snowflake for holding the data before loading it into the table.
+   - **Pipe**: Enter `emqx`, the pipe automating the loading process from the stage to the table.
+   - **Pipe User**: Enter `snowpipeuser`, the Snowflake user with the appropriate permissions to manage the pipe.
+   - **Private Key**: Enter the path to the private RSA key, for example, `file://<path to snowflake_rsa_key.private.pem>`. This is the key used for secure authentication, necessary for accessing the Snowflake pipe securely.
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
   - **Private Key**: Enter the path to the private RSA key, for example, `file://<path to snowflake_rsa_key.private.pem>`, or the content of RSA private key file. This is the key used for secure authentication, necessary for accessing the Snowflake pipe securely.
```

File: en_US/data-integration/snowflake.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+8. Select the **Upload Mode**: Currently, only `Aggregated Upload` is supported. This method groups the results of multiple rule triggers into a single file (e.g., a CSV file) and uploads it to Snowflake, reducing the number of files and improving write efficiency.
+
+9. Select **Aggregation Type**: Currently, only `csv` is supported. Data will be written to Snowflake in comma-separated CSV format.
```
Comment (id) [inline en_US/data-integration/snowflake.md]:
```suggestion
9. Select **Aggregation Type**: Currently, only `csv` is supported. Data will be staged to Snowflake in comma-separated CSV format.
```

File: en_US/data-integration/snowflake.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+### Initialize Snowflake ODBC driver
+
+To enable EMQX to communicate with Snowflake and efficiently transfer data, it is necessary to install and configure the Snowflake Open Database Connectivity (ODBC) driver. It acts as the communication bridge, ensuring that data is properly formatted, authenticated, and transferred.
+
+#### Linux
+
+Run the following script to install the Snowflake ODBC driver and configure the `odbc.ini` file:
+
+```
+scripts/install-snowflake-driver.sh
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md:62]: I wonder if it isn't better to just delegate the ODBC driver installation instructions to Snowflake docs: https://docs.snowflake.com/en/developer-guide/odbc/odbc

This script is intended for our local development and testing environment, and does some choices that are very particular for these environments.

Or at least warn the users that this script is for testing only, not a recommendation on how to setup production environments.

File: en_US/data-integration/snowflake.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+#### Generate RSA Key Pair
+
+To securely connect to Snowflake, generate an RSA key pair for authentication using the following commands:
+
+```bash
+openssl genrsa 2048 | openssl pkcs8 -topk8 -inform PEM -out snowflake_rsa_key.private.pem -nocrypt
+openssl rsa -in snowflake_rsa_key.private.pem -pubout -out snowflake_rsa_key.public.pem
+```
+
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md:139]: I suggest adding a reference to Snowflake's docs on this matter: https://docs.snowflake.com/en/user-guide/key-pair-auth

File: en_US/data-integration/snowflake.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+   ```sql
+   CREATE USER IF NOT EXISTS snowpipeuser
+       PASSWORD = 'Snowpipeuser99'
+       MUST_CHANGE_PASSWORD = FALSE;
+   
+   ALTER USER snowpipeuser SET RSA_PUBLIC_KEY = '
+   <YOUR_PUBLIC_KEY_CONTENTS_LINE_1>
+   <YOUR_PUBLIC_KEY_CONTENTS_LINE_2>
+   <YOUR_PUBLIC_KEY_CONTENTS_LINE_3>
+   <YOUR_PUBLIC_KEY_CONTENTS_LINE_4>
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md:181]: Add a comment specifying that one needs to remove the '-----BEGIN PUBLIC KEY-----' and '-----END PUBLIC KEY-----' lines from the PEM file, and include the remaining content preserving line breaks.

File: en_US/data-integration/snowflake.md [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+## Create a Connector
+
+Before adding the Snowflake Sink, you need to create the corresponding connector in EMQX to establish the connection with Snowflake.
+
+1. Go to the Dashboard **Integration** -> **Connector** page.
+2. Click the **Create** button in the top right corner.
+3. Select **Snowflake** as the connector type and click next.
+4. Enter the connector name, a combination of upper and lowercase letters and numbers. Here, enter `my-snowflake`.
+5. Enter the connection information.
+   - **Server Host**: The server host is the Snowflake endpoint URL, typically in the format `<Your Snowflake Organization ID>-<Your Snowflake Account>.snowflakecomputing.com`. You need to replace `<Your Snowflake Organization ID>-<Your Snowflake Account>` with the subdomain specific to your Snowflake instance.
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md]:
```suggestion
   - **Server Host**: The server host is the Snowflake endpoint URL, typically in the format `<Your Snowflake Organization ID>-<Your Snowflake Account Name>.snowflakecomputing.com`. You need to replace `<Your Snowflake Organization ID>-<Your Snowflake Account Name>` with the subdomain specific to your Snowflake instance.
```

File: en_US/data-integration/snowflake.md [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+   FROM
+       "t/#"
+   ```
+
+   ::: tip
+
+   If you are new to SQL, you can click **SQL Examples** and **Enable Debug** to learn and test the rule SQL results.
+
+   :::
+
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md:257]:
```suggestion
   ::: tip
   
   For Snowflake integration, it's important that the selected fields exactly match the number of columns and their names of the table defined in Snowflake, so avoid adding extra fields or selecting from `*`. 
   
   :::


```

File: en_US/data-integration/snowflake.md [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+6. Select the `my-snowflake` connector created earlier from the connector dropdown. You can also click the create button next to the dropdown to quickly create a new connector in the pop-up box. The required configuration parameters can be found in [Create a Connector](#create-a-connector).
+
+7. Configure the following settings:
+
+   - **Database Name**: Enter `testdatabase`. This is the Snowflake database that was created for storing EMQX data.
+   - **Schema**: Enter `public`, the schema within the `testdatabase` where the data table is located.
+   - **Stage**: Enter `emqx`, the stage created in Snowflake for holding the data before loading it into the table.
+   - **Pipe**: Enter `emqx`, the pipe automating the loading process from the stage to the table.
+   - **Pipe User**: Enter `snowpipeuser`, the Snowflake user with the appropriate permissions to manage the pipe.
+   - **Private Key**: Enter the path to the private RSA key, for example, `file://<path to snowflake_rsa_key.private.pem>`, or the content of RSA private key file. This is the key used for secure authentication, necessary for accessing the Snowflake pipe securely.
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md]:
```suggestion
   - **Private Key**: Enter the path to the private RSA key, for example, `file://<path to snowflake_rsa_key.private.pem>`, or the content of RSA private key file. This is the key used for secure authentication, necessary for accessing the Snowflake pipe securely.  Note that, if using a file path, the file path must be the same on all cluster nodes and must be readable by the EMQX application user.
```

File: en_US/data-integration/snowflake.md [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+12. Back on the rule creation page, click the **Create** button to complete the entire rule creation process.
+
+You have now successfully created the rule. You can see the newly created rule on the **Rules** page and the new Snowflake Sink on the **Actions (Sink)** tab.
+
+You can also click **Integration** -> **Flow Designer** to view the topology. The topology visually shows how messages under the topic `t/#` are written into the Snowflake after being parsed by the rule `my_rule`.
+
+## Test the Rule
+
+This section shows how to test the rule configured with the direct upload method.
```
Comment (thalesmg) [inline en_US/data-integration/snowflake.md]:
```suggestion
This section shows how to test the configured rule.
```

(there's no direct mode)

File: zh_CN/data-integration/snowflake.md [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+```bash
+openssl genrsa 2048 | openssl pkcs8 -topk8 -inform PEM -out snowflake_rsa_key.private.pem -nocrypt
+openssl rsa -in snowflake_rsa_key.private.pem -pubout -out snowflake_rsa_key.public.pem
+```
+
+更多信息请参考 [Key-pair authentication and key-pair rotation](https://docs.snowflake.com/en/user-guide/key-pair-auth)。
+
+#### 使用 SQL 设置 Snowflake 资源
+
+在设置 ODBC 驱动并生成 RSA 密钥对后，可以通过 SQL 命令在 Snowflake 中创建所需的数据库、表、stage 和 pipe。
```
Comment (Meggielqk) [inline zh_CN/data-integration/snowflake.md]:
```suggestion
在设置 ODBC 驱动并生成 RSA 密钥对后，可以通过 SQL 命令在 Snowflake 中创建所需的数据库、表、存储区和管道。
```

File: zh_CN/data-integration/snowflake.md [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+
+1. 在 Snowflake 控制台中，打开 SQL 工作表并执行以下 SQL 命令来创建数据库、表、stage 和 pipe：
```
Comment (Meggielqk) [inline zh_CN/data-integration/snowflake.md]:
```suggestion
1. 在 Snowflake 控制台中，打开 SQL 工作表并执行以下 SQL 命令来创建数据库、表、存储区和管道：
```

**Review by thalesmg — APPROVED** [review]

**Pull request merged**